File: 70.ts
// 70. Climbing Stairs
// https://leetcode.com/problems/climbing-stairs/


// Dynamic Programming
// time complexity: O(n) || space complexity: O(n)
function climbStairs(n: number): number {
  let result = 0;

  const dp: number[] = [];
  dp[0] = null;
  dp[1] = 1;
  dp[2] = 2;

  for (let i = 3; i < n + 1; i++) {
      dp[i] = dp[i - 1] + dp[i - 2];
  }

  result = dp[n];

  return result;
};

File: LeetCode - Top Interview 150/70.py
# 70. Climbing Stairs
# https://leetcode.com/problems/climbing-stairs/


from typing import List


class Solution:
    # 무언가 누적이라는 느낌이 들면 DP를 사용해야할 것 같네.
    # time complexity: O(n)
    def climbStairs(self, n: int) -> int:
        ways = [1, 2]

        for i in range(2, n):
            ways.append((ways[i - 2]) + (ways[i - 1]))

        return ways[n - 1]
